dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistry.java

```diff
@@ -325,13 +328,25 @@ protected void subscribeURLs(URL url, NotifyListener listener, Set<String> servi
                 serviceListeners.put(serviceNamesKey, serviceInstancesChangedListener);
             }
 
+            ApplicationModel applicationModel = url.getApplicationModel();
+            GlobalMetricsEventMulticaster eventMulticaster = applicationModel.getBeanFactory().getBean(GlobalMetricsEventMulticaster.class);
+            TimePair timePair = TimePair.start();
+
+            eventMulticaster.publishEvent(new RegistryEvent.MetricsServiceSubscribeEvent(applicationModel, timePair, serviceKey));
             if (!serviceInstancesChangedListener.isDestroyed()) {
                 listener.addServiceListener(serviceInstancesChangedListener);
                 serviceInstancesChangedListener.addListenerAndNotify(url, listener);
-                serviceDiscovery.addServiceInstancesChangedListener(serviceInstancesChangedListener);
+                try {
+                    serviceDiscovery.addServiceInstancesChangedListener(serviceInstancesChangedListener);
+                } catch (Throwable t) {
+                    eventMulticaster.publishErrorEvent(new RegistryEvent.MetricsServiceSubscribeEvent(applicationModel, timePair, serviceKey));
+                    throw t;
+                }
+                eventMulticaster.publishFinishEvent(new RegistryEvent.MetricsServiceSubscribeEvent(applicationModel, timePair, serviceKey));
```

Comment:
Can we exact these logic into separated method like `doExecute(Runnable)`

Reply:
Will be optimized together later